src/ecs/systems/audio.rs

```diff
@@ -0,0 +1,117 @@
+use std::iter::Iterator;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::time::Duration;
+
+use cgmath::{Matrix4, Transform, Point3};
+use ecs::{Join, System, ReadStorage, WriteStorage, Entity};
+use ecs::components::{AudioEmitter, AudioListener};
+use ecs::components::Transform as TransformComponent;
+use rodio::{SpatialSink, Source, Sample};
+
+/// Syncs 3D transform data with the audio engine to provide 3D audio.
+pub struct AudioSystem {
+    listener: Entity,
+}
+
+// Wraps a source and signals to another thread when that source has ended.
+struct EndSignalSource<I: Source>
+where
+    <I as Iterator>::Item: Sample,
+{
+    input: I,
+    signal: Arc<AtomicBool>,
+}
+
+impl<I: Source> EndSignalSource<I>
+where
+    <I as Iterator>::Item: Sample,
+{
+    pub fn new(input: I, signal: Arc<AtomicBool>) -> EndSignalSource<I> {
+        EndSignalSource { input, signal }
+    }
+}
+
+impl<I: Source> Iterator for EndSignalSource<I>
+where
+    <I as Iterator>::Item: Sample,
+{
+    type Item = <I as Iterator>::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let next = self.input.next();
+        if next.is_none() {
+            self.signal.store(true, Ordering::Relaxed);
+        }
+        next
+    }
+}
+
+impl<I: Source> Source for EndSignalSource<I>
+where
+    <I as Iterator>::Item: Sample,
+{
+    fn current_frame_len(&self) -> Option<usize> {
+        self.input.current_frame_len()
+    }
+
+    fn channels(&self) -> u16 {
+        self.input.channels()
+    }
+
+    fn samples_rate(&self) -> u32 {
+        self.input.samples_rate()
+    }
+
+    fn total_duration(&self) -> Option<Duration> {
+        self.input.total_duration()
+    }
+}
+
+impl<'a> System<'a> for AudioSystem {
+    type SystemData = (ReadStorage<'a, TransformComponent>,
+     ReadStorage<'a, AudioListener>,
+     WriteStorage<'a, AudioEmitter>);
+
+    fn run(&mut self, (transform, listener, mut audio_emitter): Self::SystemData) {
+        if let Some(listener) = listener.get(self.listener) {
+            if let Some(listener_transform) = transform.get(self.listener) {
+                let listener_transform = Matrix4::from(listener_transform.0);
+                let left_ear_position = listener_transform
+                    .transform_point(Point3::from(listener.left_ear))
+                    .into();
+                let right_ear_position = listener_transform
+                    .transform_point(Point3::from(listener.right_ear))
+                    .into();
+                for (transform, mut audio_emitter) in (&transform, &mut audio_emitter).join() {
+                    let x = transform.0[3][0];
+                    let y = transform.0[3][1];
+                    let z = transform.0[3][2];
+                    let emitter_position = [x, y, z];
+                    // Remove all sinks whose sounds have ended.
+                    for i in 0..audio_emitter.sinks.len() {
```

Comment:
This is broken and will cause panics.  I was trying to emulate the `retain` functionality here since that doesn't exist on SmallVec.  I've opened a PR to add `retain` to SmallVec [here](https://github.com/servo/rust-smallvec/pull/59)